File: frontend/src/types/asset.ts
/**
 * Asset domain types. Catalog values are stored in English (they match the
 * backend `StrEnum`s exactly); the UI translates them only for display.
 */

export const ASSET_TYPES = [
  "Server",
  "Virtual Machine",
  "Database",
  "Application",
  "Network Device",
  "Container",
  "Kubernetes Cluster",
  "Cloud Resource",
] as const;
export type AssetType = (typeof ASSET_TYPES)[number];

export const ENVIRONMENTS = ["Production", "Staging", "Development", "Test"] as const;
export type Environment = (typeof ENVIRONMENTS)[number];

export const CRITICALITIES = ["Critical", "High", "Medium", "Low"] as const;
export type Criticality = (typeof CRITICALITIES)[number];

export const ASSET_STATUSES = ["Operational", "Degraded", "Maintenance", "Offline"] as const;
export type AssetStatus = (typeof ASSET_STATUSES)[number];

export interface Asset {
  id: string;
  name: string;
  asset_type: AssetType;
  environment: Environment;
  criticality: Criticality;
  status: AssetStatus;
  hostname: string | null;
  ip_address: string | null;
  owner: string | null;
  description: string | null;
  is_active: boolean;
  created_at: string;
  updated_at: string;
}

export interface AssetPage {
  items: Asset[];
  page: number;
  page_size: number;
  total: number;
  total_pages: number;
}

/** Fields the create form submits. */
export interface AssetCreateInput {
  name: string;
  asset_type: AssetType;
  environment: Environment;
  criticality: Criticality;
  status: AssetStatus;
  hostname?: string | null;
  ip_address?: string | null;
  owner?: string | null;
  description?: string | null;
}

/** Partial update - any subset of the content fields (never `is_active`). */
export type AssetUpdateInput = Partial<AssetCreateInput>;

function isOneOf<T extends readonly string[]>(list: T, v: unknown): v is T[number] {
  return typeof v === "string" && (list as readonly string[]).includes(v);
}

/** The network response is untrusted - validate its shape. */
export function isAsset(value: unknown): value is Asset {
  if (typeof value !== "object" || value === null) return false;
  const v = value as Record<string, unknown>;
  return (
    typeof v.id === "string" &&
    typeof v.name === "string" &&
    isOneOf(ASSET_TYPES, v.asset_type) &&
    isOneOf(ENVIRONMENTS, v.environment) &&
    isOneOf(CRITICALITIES, v.criticality) &&
    isOneOf(ASSET_STATUSES, v.status) &&
    (v.hostname === null || typeof v.hostname === "string") &&
    (v.ip_address === null || typeof v.ip_address === "string") &&
    (v.owner === null || typeof v.owner === "string") &&
    (v.description === null || typeof v.description === "string") &&
    typeof v.is_active === "boolean" &&
    typeof v.created_at === "string" &&
    typeof v.updated_at === "string"
  );
}

/** Aggregate inventory counts from `GET /api/v1/assets/summary`. */
export interface AssetSummary {
  total: number;
  active: number;
  inactive: number;
  by_criticality: Record<string, number>;
  by_status: Record<string, number>;
  by_environment: Record<string, number>;
  by_type: Record<string, number>;
}

function isCountMap(v: unknown): v is Record<string, number> {
  return (
    typeof v === "object" &&
    v !== null &&
    Object.values(v as Record<string, unknown>).every((n) => typeof n === "number")
  );
}

export function isAssetSummary(value: unknown): value is AssetSummary {
  if (typeof value !== "object" || value === null) return false;
  const v = value as Record<string, unknown>;
  return (
    typeof v.total === "number" &&
    typeof v.active === "number" &&
    typeof v.inactive === "number" &&
    isCountMap(v.by_criticality) &&
    isCountMap(v.by_status) &&
    isCountMap(v.by_environment) &&
    isCountMap(v.by_type)
  );
}

export function isAssetPage(value: unknown): value is AssetPage {
  if (typeof value !== "object" || value === null) return false;
  const v = value as Record<string, unknown>;
  return (
    Array.isArray(v.items) &&
    v.items.every(isAsset) &&
    typeof v.page === "number" &&
    typeof v.page_size === "number" &&
    typeof v.total === "number" &&
    typeof v.total_pages === "number"
  );
}

File: backend/app/models/asset.py
"""The ``Asset`` entity - InfraGuard's first business-domain model.

An asset is any tracked piece of infrastructure: a server, VM, database,
application, network device, container, Kubernetes cluster or cloud resource.

Deliberately flat for this milestone: identity + classification + a few network
details + lifecycle. Dependencies, incidents, health telemetry and obsolescence
are later phases and are **not** modelled here.

Catalog values (``asset_type``, ``environment``, ``criticality``, ``status``) are
small, stable enumerations. They are stored as their English string value and
constrained by a database ``CHECK`` built from the ``StrEnum`` below - no
separate catalog tables (overkill for a fixed vocabulary) and no native
PostgreSQL ``ENUM`` type (which needs a migration to extend). Display
translation happens in the frontend.
"""

from __future__ import annotations

import enum
import uuid
from datetime import datetime

from sqlalchemy import (
    Boolean,
    CheckConstraint,
    DateTime,
    ForeignKey,
    Index,
    String,
    Text,
    func,
    text,
)
from sqlalchemy.dialects.postgresql import UUID as PGUUID
from sqlalchemy.orm import Mapped, mapped_column

from app.db.base import Base


class AssetType(enum.StrEnum):
    SERVER = "Server"
    VIRTUAL_MACHINE = "Virtual Machine"
    DATABASE = "Database"
    APPLICATION = "Application"
    NETWORK_DEVICE = "Network Device"
    CONTAINER = "Container"
    KUBERNETES_CLUSTER = "Kubernetes Cluster"
    CLOUD_RESOURCE = "Cloud Resource"


class Environment(enum.StrEnum):
    PRODUCTION = "Production"
    STAGING = "Staging"
    DEVELOPMENT = "Development"
    TEST = "Test"


class Criticality(enum.StrEnum):
    CRITICAL = "Critical"
    HIGH = "High"
    MEDIUM = "Medium"
    LOW = "Low"


class AssetStatus(enum.StrEnum):
    OPERATIONAL = "Operational"
    DEGRADED = "Degraded"
    MAINTENANCE = "Maintenance"
    OFFLINE = "Offline"


# Field length bounds - also mirrored by the Pydantic schemas.
NAME_MAX_LENGTH = 200
HOSTNAME_MAX_LENGTH = 253  # RFC 1035 maximum domain name length
IP_ADDRESS_MAX_LENGTH = 45  # longest possible IPv6 textual form
OWNER_MAX_LENGTH = 200
DESCRIPTION_MAX_LENGTH = 2000


def _in_check(column: str, enum_cls: type[enum.StrEnum], name: str) -> CheckConstraint:
    """Build ``CHECK (<column> IN ('A', 'B', ...))`` from an enum's values."""
    values = [member.value for member in enum_cls]
    assert all("'" not in v for v in values), "catalog values must not contain quotes"
    rendered = ", ".join(f"'{v}'" for v in values)
    return CheckConstraint(f"{column} IN ({rendered})", name=name)


class Asset(Base):
    __tablename__ = "assets"
    __table_args__ = (
        CheckConstraint("char_length(name) > 0", name="name_not_empty"),
        _in_check("asset_type", AssetType, "asset_type_valid"),
        _in_check("environment", Environment, "environment_valid"),
        _in_check("criticality", Criticality, "criticality_valid"),
        _in_check("status", AssetStatus, "status_valid"),
        Index("ix_assets_name", "name"),
        Index("ix_assets_asset_type", "asset_type"),
        Index("ix_assets_environment", "environment"),
        Index("ix_assets_criticality", "criticality"),
        Index("ix_assets_status", "status"),
        Index("ix_assets_is_active", "is_active"),
        Index("ix_assets_created_at", "created_at"),
        # Every normal query filters ``deleted_at IS NULL``; Trash filters the
        # opposite. A partial index keeps the common "active" scan cheap.
        Index(
            "ix_assets_deleted_at",
            "deleted_at",
            postgresql_where=text("deleted_at IS NOT NULL"),
        ),
    )

    id: Mapped[uuid.UUID] = mapped_column(
        PGUUID(as_uuid=True),
        primary_key=True,
        default=uuid.uuid4,
        server_default=text("gen_random_uuid()"),
    )

    name: Mapped[str] = mapped_column(String(NAME_MAX_LENGTH), nullable=False)
    asset_type: Mapped[str] = mapped_column(String(40), nullable=False)
    environment: Mapped[str] = mapped_column(String(20), nullable=False)
    criticality: Mapped[str] = mapped_column(String(10), nullable=False)
    status: Mapped[str] = mapped_column(String(20), nullable=False)

    hostname: Mapped[str | None] = mapped_column(String(HOSTNAME_MAX_LENGTH), nullable=True)
    ip_address: Mapped[str | None] = mapped_column(String(IP_ADDRESS_MAX_LENGTH), nullable=True)
    description: Mapped[str | None] = mapped_column(Text, nullable=True)
    owner: Mapped[str | None] = mapped_column(String(OWNER_MAX_LENGTH), nullable=True)

    is_active: Mapped[bool] = mapped_column(
        Boolean, nullable=False, default=True, server_default=text("true")
    )

    # --- Soft delete (Trash) -------------------------------------------------
    # ``deleted_at IS NULL`` -> live record; a timestamp -> moved to Trash and
    # excluded from every normal query / summary / picker. Restorable with the
    # same id and full history. ``deleted_by`` is a snapshot FK (SET NULL if the
    # user is later removed); the actor is always the authenticated session,
    # never a request-body value.
    deleted_at: Mapped[datetime | None] = mapped_column(
        DateTime(timezone=True), nullable=True
    )
    deleted_by: Mapped[uuid.UUID | None] = mapped_column(
        PGUUID(as_uuid=True),
        ForeignKey("users.id", ondelete="SET NULL"),
        nullable=True,
    )

    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), nullable=False, server_default=func.now()
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        nullable=False,
        server_default=func.now(),
        onupdate=func.now(),
    )

    def __repr__(self) -> str:  # pragma: no cover - debug aid
        return f"<Asset id={self.id!s} name={self.name!r} active={self.is_active}>"
